File: admin/src/calibFingerprint.ts
/** Канонический фингерпринт калибровки камеры (зеркало app/global_id/calib_fingerprint.py). */

const FNV_OFFSET = 2166136261;
const FNV_PRIME = 16777619;
export const DEFAULT_PERSON_H_M = 1.7;

function fmt(v: unknown): string {
  const n = Number(v);
  return Number.isFinite(n) ? n.toFixed(6) : "0.000000";
}

function fmtPt(pt: unknown): string {
  if (!Array.isArray(pt) || pt.length < 2) return "0.000000,0.000000";
  return `${fmt(pt[0])},${fmt(pt[1])}`;
}

function fmtSize(size: unknown): string {
  if (!Array.isArray(size) || size.length < 2) return "0,0";
  const w = Number(size[0]);
  const h = Number(size[1]);
  if (!Number.isFinite(w) || !Number.isFinite(h)) return "0,0";
  return `${Math.trunc(w)},${Math.trunc(h)}`;
}

export function canonicalCalibString(opts: {
  cameraKey: string;
  cameraDoc: Record<string, unknown> | null | undefined;
  torsoHeightM: number;
  trackingSize: [number, number] | number[] | null | undefined;
  personHeightM?: number;
}): string {
  const doc = opts.cameraDoc && typeof opts.cameraDoc === "object" ? opts.cameraDoc : {};
  const plRaw = doc.placement;
  const pl = plRaw && typeof plRaw === "object" ? (plRaw as Record<string, unknown>) : {};
  const pos = Array.isArray(pl.position) ? pl.position : [0, 0];
  const personH = opts.personHeightM ?? DEFAULT_PERSON_H_M;
  const lines = [
    "v2",
    `camera_key=${opts.cameraKey}`,
    `image_size=${fmtSize(doc.image_size)}`,
    `tracking_size=${fmtSize(opts.trackingSize)}`,
    `torso_height_m=${fmt(opts.torsoHeightM)}`,
    `person_height_m=${fmt(personH)}`,
    "placement=" +
      [fmtPt(pos), fmt(pl.yaw_deg ?? 0), fmt(pl.fov_deg ?? 70), fmt(pl.height_m ?? 3), fmt(pl.pitch_deg ?? 35)].join(
        "|",
      ),
  ];
  const h = doc.H;
  if (Array.isArray(h) && h.length >= 9) {
    lines.push("H=" + h.slice(0, 9).map(fmt).join(","));
  } else {
    lines.push("H=");
  }
  return lines.join("\n");
}

export function fnv1a32(text: string): string {
  let h = FNV_OFFSET >>> 0;
  const bytes = new TextEncoder().encode(text);
  for (const b of bytes) {
    h ^= b;
    h = Math.imul(h, FNV_PRIME) >>> 0;
  }
  return h.toString(16).padStart(8, "0");
}

export function calibFingerprint(opts: {
  cameraKey: string;
  cameraDoc: Record<string, unknown> | null | undefined;
  torsoHeightM: number;
  trackingSize: [number, number] | number[] | null | undefined;
  personHeightM?: number;
}): string {
  return fnv1a32(canonicalCalibString(opts));
}

File: app/global_id/calib_fingerprint.py
"""Канонический фингерпринт калибровки камеры (зеркало admin/src/calibFingerprint.ts)."""

from __future__ import annotations

from typing import Any

FNV_OFFSET = 2166136261
FNV_PRIME = 16777619
DEFAULT_PERSON_H_M = 1.70


def _fmt(v: Any) -> str:
    try:
        return f"{float(v):.6f}"
    except (TypeError, ValueError):
        return "0.000000"


def _fmt_pt(pt: Any) -> str:
    if not isinstance(pt, (list, tuple)) or len(pt) < 2:
        return "0.000000,0.000000"
    return f"{_fmt(pt[0])},{_fmt(pt[1])}"


def _fmt_size(size: Any) -> str:
    if not isinstance(size, (list, tuple)) or len(size) < 2:
        return "0,0"
    try:
        return f"{int(size[0])},{int(size[1])}"
    except (TypeError, ValueError):
        return "0,0"


def canonical_calib_string(
    camera_key: str,
    camera_doc: dict[str, Any] | None,
    torso_height_m: float,
    tracking_size: tuple[int, int] | list[int] | None,
    person_height_m: float = DEFAULT_PERSON_H_M,
) -> str:
    doc = camera_doc if isinstance(camera_doc, dict) else {}
    pl = doc.get("placement") if isinstance(doc.get("placement"), dict) else {}
    pos = pl.get("position") if isinstance(pl.get("position"), (list, tuple)) else (0, 0)
    lines = [
        "v2",
        f"camera_key={camera_key}",
        f"image_size={_fmt_size(doc.get('image_size'))}",
        f"tracking_size={_fmt_size(tracking_size)}",
        f"torso_height_m={_fmt(torso_height_m)}",
        f"person_height_m={_fmt(person_height_m)}",
        "placement="
        + "|".join(
            [
                _fmt_pt(pos),
                _fmt(pl.get("yaw_deg") or 0),
                _fmt(pl.get("fov_deg") or 70),
                _fmt(pl.get("height_m") if pl.get("height_m") is not None else 3),
                _fmt(pl.get("pitch_deg") if pl.get("pitch_deg") is not None else 35),
            ]
        ),
    ]
    h = doc.get("H")
    if isinstance(h, (list, tuple)) and len(h) >= 9:
        lines.append("H=" + ",".join(_fmt(x) for x in h[:9]))
    else:
        lines.append("H=")
    return "\n".join(lines)


def fnv1a32(text: str) -> str:
    h = FNV_OFFSET
    for b in text.encode("utf-8"):
        h ^= b
        h = (h * FNV_PRIME) & 0xFFFFFFFF
    return f"{h:08x}"


def calib_fingerprint(
    camera_key: str,
    camera_doc: dict[str, Any] | None,
    torso_height_m: float,
    tracking_size: tuple[int, int] | list[int] | None,
    person_height_m: float = DEFAULT_PERSON_H_M,
) -> str:
    return fnv1a32(
        canonical_calib_string(camera_key, camera_doc, torso_height_m, tracking_size, person_height_m)
    )
